### DISCUSSION TITLE
Fetching Dynamic MP3 URLs for AudioSourceURL

### DISCUSSION THREAD

**Discussion by abramadams** [pschatzmann/arduino-audio-tools#1263] [Q&A] [answered]
Hello!  First off, what an awesome library.  It has been extremely helpful to help me learn (new to c++, esp32 and working with audio).

I'm trying to adapt the https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-url-i2s/ example to be able to fetch an array of urls instead of having them statically defined.  The goal is to hit an https endpoint to fetch an array of 2-5 mp3 URLs, play through them, then fetch more, etc...  The example works great on my board setup, which is great, I just need to find a way to make the array of urls dynamic.

I can easily hit my webserver and retrieve the array of urls, however it seems the AudioSourceURL constructor expects the urls to be defined once globally, and there aren't methods on the class to update or set them after instantiation.  Is `AudioSourceURL` not the correct class for what I'm attempting?  As I mentioned, I'm new to c++ and am using the Arduino IDE so not sure if/how to extend classes (if that is what is needed).

If there are other examples that would help put me on the right track, or different methods to accomplish this I would be most grateful for any direction anyone can provide.

Thanks in advance!

**Comment by pschatzmann** [comment, marked as answer]
You will find the answer in the [documentation](https://github.com/pschatzmann/arduino-audio-tools/wiki/The-Audio-Player-Class): An [AudioSource](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_source.html) is the abstract API used by the AudioPlayer class to navigate and access the audio stream. __You can create your own source__ or just use one of the existing implementations.

Just look at the current implementation of AudioSourceURL and replace the c array of char* with a ```Vector<String>``` or ```std::vector<String>``` to create your own AudioSourceDynamicURL class.

The existing AudioSourceURL tries to be as memory efficient as possible by using the data stored in Progmem where the data can not be changed. This prevents a dynamic solution!

**Comment by pschatzmann** [comment]
I have added the new class [AudioSourceDynamicURL](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_source_dynamic_u_r_l.html)

**Comment by abramadams** [comment]
Thank you @pschatzmann!! Your tip about replacing the c array of char* with a Vector<String> helped get me on the right track.  I ended up creating a subclass last night that worked perfectly, but didn't reply yet as I was still troubleshooting some performance issues (turned out to be an issue with the bitrate of my mp3 files).  Then to have you add that AudioSourceDynamicURL class (which is much more elegant than mine) was a very nice surprise this morning!
